packages/components/src/number-control/index.tsx

```diff
@@ -105,10 +118,13 @@ export function NumberControl(
 		 * Handles drag to update events
 		 */
 		if ( type === inputControlActionTypes.DRAG && isDragEnabled ) {
+			// @ts-expect-error TODO: Investigate
 			const [ x, y ] = payload.delta;
+			// @ts-expect-error TODO: Investigate
 			const enableShift = payload.shiftKey && isShiftStepEnabled;
```

Comment:
These two lines do not error in my IDE, but error when I do `tsc --build` by CLI. Just me? 🤔

Reply:
Actually, it looks like in #38753 I had solved the situation by [casting `action` to the `DragAction` type](https://github.com/WordPress/gutenberg/pull/38753/files#diff-1795d70ae6f25ef5ee79fa75d215a18e87de7d7bdc577b294376838325d983a8R114), which seems like a good solution.

The only thing that will still need to be fixed is that, even in that case, `payload` doesn't seem to have a `delta` prop, which means that:
- either our [`DragProps` types](https://github.com/WordPress/gutenberg/blob/8f6edb18656f150fec6a77f4005397573796a264/packages/components/src/input-control/types.ts) aren't exhaustive
- or somewhat the `delta` prop is just not there at runtime too (we should check 😅 )

Reply:
> Same for me. Maybe it's a mismatch between the TS version?

It wasn't for me. I even cleared the TS cache. Today, only the second line errors in my IDE. Very weird and... unsettling! 🙁

So, in my IDE, `if ( type === inputControlActionTypes.DRAG )` can successfully narrow down the action type to a `DragAction`, which semi-correctly changes the `payload` type. So I think it is theoretically possible to type the Actions correctly, and rely on type narrowing, so we don't have to typecast all over the place in the consumers.

I'd prefer to bump this to a separate task, so we can look into it better. It does look possible to type the [events](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/input-control/reducer/actions.ts) correctly with generics. How does that sound?

Reply:
Agreed, let's look into this as a separate task — that way we can tackle all drag-related types

Reply:
Moved to #43945